tests/suites/test_suite_aes.function

```diff
@@ -151,6 +151,157 @@ exit:
 }
 /* END_CASE */
 
+/* BEGIN_CASE depends_on:MBEDTLS_CIPHER_MODE_XTS */
+void aes_encrypt_xex( char *hex_key_string, char *hex_iv_string,
+                      char *hex_src_string, char *hex_dst_string )
+{
+    unsigned char key_str[100] = { 0, };
+    unsigned char iv_str[100]  = { 0, };
+    unsigned char src_str[100] = { 0, };
+    unsigned char dst_str[100] = { 0, };
+    unsigned char output[100]  = { 0, };
+    mbedtls_aes_xts_context ctx;
+    int key_len, data_len;
+
+    mbedtls_aes_xts_init( &ctx );
+
+    key_len = unhexify( key_str, hex_key_string );
+    unhexify( iv_str, hex_iv_string );
+    data_len = unhexify( src_str, hex_src_string );
+
+    mbedtls_aes_xts_setkey_enc( &ctx, key_str, key_len * 8 );
+
+    TEST_ASSERT( mbedtls_aes_crypt_xex( &ctx, MBEDTLS_AES_ENCRYPT, data_len, iv_str, src_str, output ) == 0);
+    hexify( dst_str, output, data_len );
+    TEST_ASSERT( strcmp( (char *) dst_str, hex_dst_string ) == 0 );
+
+exit:
+    mbedtls_aes_xts_free( &ctx );
+}
+/* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_CIPHER_MODE_XTS */
+void aes_decrypt_xex( char *hex_key_string, char *hex_iv_string,
+                      char *hex_src_string, char *hex_dst_string )
+{
+    unsigned char key_str[100] = { 0, };
+    unsigned char iv_str[100]  = { 0, };
+    unsigned char src_str[100] = { 0, };
+    unsigned char dst_str[100] = { 0, };
+    unsigned char output[100]  = { 0, };
+    mbedtls_aes_xts_context ctx;
+    int key_len, data_len;
+
+    mbedtls_aes_xts_init( &ctx );
+
+    key_len = unhexify( key_str, hex_key_string );
+    unhexify( iv_str, hex_iv_string );
+    data_len = unhexify( src_str, hex_src_string );
+
+    mbedtls_aes_xts_setkey_dec( &ctx, key_str, key_len * 8 );
+
+    TEST_ASSERT( mbedtls_aes_crypt_xex( &ctx, MBEDTLS_AES_DECRYPT, data_len, iv_str, src_str, output ) == 0 );
+
+    hexify( dst_str, output, data_len );
+    TEST_ASSERT( strcmp( (char *) dst_str, hex_dst_string ) == 0 );
+
+exit:
+    mbedtls_aes_xts_free( &ctx );
+}
+/* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_CIPHER_MODE_XTS */
+void aes_encrypt_xts( char *hex_key_string, char *hex_data_unit_string,
+                      char *hex_src_string, char *hex_dst_string )
+{
+    enum { AES_BLOCK_SIZE = 16 };
+    unsigned char *key;
+    unsigned char *data_unit;
+    unsigned char *src;
+    unsigned char *dst;
+    unsigned char *output;
+    mbedtls_aes_xts_context ctx;
+    size_t key_len, src_len, dst_len, data_unit_len;
+
+    mbedtls_aes_xts_init( &ctx );
+
+    data_unit = unhexify_alloc( hex_data_unit_string, &data_unit_len );
+    TEST_ASSERT( data_unit_len == AES_BLOCK_SIZE );
+
+    key = unhexify_alloc( hex_key_string, &key_len );
+    TEST_ASSERT( key_len % 2 == 0 );
+
+    src = unhexify_alloc( hex_src_string, &src_len );
+    dst = unhexify_alloc( hex_dst_string, &dst_len );
+    TEST_ASSERT( src_len == dst_len );
```

Comment:
If this fails then the test vector is wrong not the library. Do we need to test this?

Reply:
I would be worried about having improperly described tests generate results which claim to pass. If lucky these failures would mean the library could only work with itself, but it is possible (though very unlikely) that a weakness could generate code compatible with other implementations that is weaker in some way. The biggest concern would be key generation, which this is not.

So in this case the test might be redundant. In general I am not sure it is a good idea to discourage test sanity checks.

Reply:
I have this test present to catch typos in the test vectors file. For instance, if you copy paste everything but the last little bit of the `src`, we'd catch that here.